In [93]:
import pandas as pd

In [94]:
reviews = pd.DataFrame(pd.read_pickle('../data/reviews.pkl'))
reviews.head()


product_url  \
0                                     https://www.rossmann.pl/Produkt/Balsamy-i-olejki-do-brody/Rituals-Homme-balsam-do-stylizacji-brody-30-ml,2106072,13086   
1  https://www.rossmann.pl/Produkt/Zestawy-zapachowe/Jaguar-Era-Reloaded-zestaw-dla-mezczyzn-woda-toaletowa-100ml-zel-pod-prysznic-200ml-1-szt,2083791,13263   
2  https://www.rossmann.pl/Produkt/Zestawy-zapachowe/Jaguar-Era-Reloaded-zestaw-dla-mezczyzn-woda-toaletowa-100ml-zel-pod-prysznic-200ml-1-szt,2083791,13263   
3                                     https://www.rossmann.pl/Produkt/Wody-toaletowe/Hugo-Boss-Just-Different-woda-toaletowa-dla-mezczyzn-75-ml,235165,13274   
4                                     https://www.rossmann.pl/Produkt/Wody-toaletowe/Hugo-Boss-Just-Different-woda-toaletowa-dla-mezczyzn-75-ml,235165,13274   

   rating      author        date                                    content  
0     5.0         EwA  25.06.2025  Mąż bardzo zadowolony z jakości i zapachu  
1     5.0       Olena  21.05.2025                                             
2     4.0    Wojciech  21.05.2025                           przyjemny zapach  
3     5.0       Daria  22.06.2025                                             
4     5.0  Aleksandra  12.06.2025

In [95]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10348 entries, 0 to 10347
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   product_url  10348 non-null  object 
 1   rating       10334 non-null  float64
 2   author       10348 non-null  object 
 3   date         10348 non-null  object 
 4   content      10348 non-null  object 
dtypes: float64(1), object(4)
memory usage: 404.3+ KB


In [96]:
# 14 wpisów nie posiadających ratingu
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)
pd.set_option('display.width', 0)
missing_data_url_list = list(reviews[reviews['rating'].isnull()]['product_url'].unique())
missing_data_url_list

['https://www.rossmann.pl/Produkt/Wody-toaletowe/Azzaro-Chrome-Legend-woda-toaletowa-dla-mezczyzn-125-ml,168929,13274',
 'https://www.rossmann.pl/Produkt/Minikosmetyki/Isana-Men-zel-pod-prysznic-3w1-mini-Fresh-50-ml,163253,19247',
 'https://www.rossmann.pl/Produkt/Szampony-i-odzywki/Lbiotica-Biovax-Men-Anti-Hair-Loss-Volume-Boost-szampon-do-wlosow-i-zarostu-przeciw-wypadaniu-wlos,2083075,13253',
 'https://www.rossmann.pl/Produkt/Spraye-i-mgielki-do-wlosow/Farmona-Radical-mgielka-do-wlosow-oslabionych-i-wypadajacych-wzmacniajaca-200-ml,279857,13184',
 'https://www.rossmann.pl/Produkt/Antyperspiranty/Nivea-Men-Black-White-Invisible-antyperspirant-w-sprayu-72h-Ultimate-Impact-150-ml,342238,13120',
 'https://www.rossmann.pl/Produkt/Wody-po-goleniu/Gillette-Series-Arctic-Ice-woda-po-goleniu-dla-mezczyzn-100-ml,13646,13244',
 'https://www.rossmann.pl/Produkt/Blokery-potu/Ziaja-Yego-Bloker-antyperspirant-w-kulce-dla-mezczyzn-60-ml,173254,13121',
 'https://www.rossmann.pl/Produkt/Zele-i-pianki

Powyższe produkty posiadają opinie i oceny. Ich brak może wynikać z błędów powstałych w trakcie pozyskiwania danych. W celu ich pozyskania ponownie przeprowadzę pozyskiwanie danych tylko dla tych linków.

In [97]:
pd.to_pickle(missing_data_url_list, '../data/missing_data.pkl')

Przy użyciu lekko zmodyfikowanego scrapera () ponownie pobrałem dane ocen użytkowników dla linków z brakujacymi danymi.

In [98]:
reviews_missing = pd.DataFrame(pd.read_pickle('../data/reviews_missing.pkl'))
reviews_missing.head(20)

product_url  \
0                                                   https://www.rossmann.pl/Produkt/Wody-toaletowe/Azzaro-Chrome-Legend-woda-toaletowa-dla-mezczyzn-125-ml,168929,13274   
1                                                   https://www.rossmann.pl/Produkt/Wody-toaletowe/Azzaro-Chrome-Legend-woda-toaletowa-dla-mezczyzn-125-ml,168929,13274   
2                                                   https://www.rossmann.pl/Produkt/Wody-toaletowe/Azzaro-Chrome-Legend-woda-toaletowa-dla-mezczyzn-125-ml,168929,13274   
3                                                            https://www.rossmann.pl/Produkt/Minikosmetyki/Isana-Men-zel-pod-prysznic-3w1-mini-Fresh-50-ml,163253,19247   
4                                                            https://www.rossmann.pl/Produkt/Minikosmetyki/Isana-Men-zel-pod-prysznic-3w1-mini-Fresh-50-ml,163253,19247   
5                                                            https://www.rossmann.pl/Produkt/Minikosmetyki/Isana-Men-zel-pod-prysznic-3w1-mini-Fresh-50-ml,163253,19247   
6                                                            https://www.rossmann.pl/Produkt/Minikosmetyki/Isana-Men-zel-pod-prysznic-3w1-mini-Fresh-50-ml,163253,19247   
7                                                            https://www.rossmann.pl/Produkt/Minikosmetyki/Isana-Men-zel-pod-prysznic-3w1-mini-Fresh-50-ml,163253,19247   
8                                                            https://www.rossmann.pl/Produkt/Minikosmetyki/Isana-Men-zel-pod-prysznic-3w1-mini-Fresh-50-ml,163253,19247   
9                                                            https://www.rossmann.pl/Produkt/Minikosmetyki/Isana-Men-zel-pod-prysznic-3w1-mini-Fresh-50-ml,163253,19247   
10                                                           https://www.rossmann.pl/Produkt/Minikosmetyki/Isana-Men-zel-pod-prysznic-3w1-mini-Fresh-50-ml,163253,19247   
11                                                           https://www.rossmann.pl/Produkt/Minikosmetyki/Isana-Men-zel-pod-prysznic-3w1-mini-Fresh-50-ml,163253,19247   
12                                                           https://www.rossmann.pl/Produkt/Minikosmetyki/Isana-Men-zel-pod-prysznic-3w1-mini-Fresh-50-ml,163253,19247   
13                                                           https://www.rossmann.pl/Produkt/Minikosmetyki/Isana-Men-zel-pod-prysznic-3w1-mini-Fresh-50-ml,163253,19247   
14                                                           https://www.rossmann.pl/Produkt/Minikosmetyki/Isana-Men-zel-pod-prysznic-3w1-mini-Fresh-50-ml,163253,19247   
15                                                           https://www.rossmann.pl/Produkt/Minikosmetyki/Isana-Men-zel-pod-prysznic-3w1-mini-Fresh-50-ml,163253,19247   
16                                                           https://www.rossmann.pl/Produkt/Minikosmetyki/Isana-Men-zel-pod-prysznic-3w1-mini-Fresh-50-ml,163253,19247   
17  https://www.rossmann.pl/Produkt/Szampony-i-odzywki/Lbiotica-Biovax-Men-Anti-Hair-Loss-Volume-Boost-szampon-do-wlosow-i-zarostu-przeciw-wypadaniu-wlos,2083075,13253   
18  https://www.rossmann.pl/Produkt/Szampony-i-odzywki/Lbiotica-Biovax-Men-Anti-Hair-Loss-Volume-Boost-szampon-do-wlosow-i-zarostu-przeciw-wypadaniu-wlos,2083075,13253   
19  https://www.rossmann.pl/Produkt/Szampony-i-odzywki/Lbiotica-Biovax-Men-Anti-Hair-Loss-Volume-Boost-szampon-do-wlosow-i-zarostu-przeciw-wypadaniu-wlos,2083075,13253   

    rating      author        date  \
0      4.0      Maryla  17.06.2025   
1      5.0     Mariusz  14.06.2025   
2      NaN                           
3      5.0      Oliwia  26.06.2025   
4      5.0     Justyna  23.06.2025   
5      4.0        Igor  11.06.2025   
6      5.0   Kazimierz  07.06.2025   
7      5.0     Klaudia  05.06.2025   
8      5.0    Krzysiek  31.05.2025   
9      5.0     Andrzej  28.05.2025   
10     5.0        Jola  27.05.2025   
11     5.0    Wiktoria  22.05.2025   
12     5.0     Justyna  22.05.2025   
13     5.0      Alicja  18.05.20

Po przeanalizowaniu powyższych danych i stron produktów odkryłem przyczynę brakujących danych - brakujące dane dotyczą komentarzy, które zostały usunięte ze strony. Mogę więc bezpiecznie usunąć wpisy z pierwszego data frame.

In [99]:
reviews = reviews[reviews['rating'].notnull()]
reviews.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10334 entries, 0 to 10347
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   product_url  10334 non-null  object 
 1   rating       10334 non-null  float64
 2   author       10334 non-null  object 
 3   date         10334 non-null  object 
 4   content      10334 non-null  object 
dtypes: float64(1), object(4)
memory usage: 484.4+ KB


Liczba unikalnych produktów, dla których posiadamy dane.

In [100]:
reviews['product_url'].nunique()

856

Porównanie z danymi produktów po etapie EDA.

In [101]:
products = pd.read_pickle('../data/male_data_after_eda.pkl')
products.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1482 entries, 0 to 1793
Data columns (total 54 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   product_id                         1482 non-null   int64  
 1   name                               1482 non-null   string 
 2   caption                            1482 non-null   string 
 3   cat_1                              1482 non-null   string 
 4   cat_2                              1482 non-null   string 
 5   cat_3                              1482 non-null   string 
 6   cat_4                              382 non-null    string 
 7   cat_5                              1482 non-null   string 
 8   url                                1482 non-null   string 
 9   current_price_pln                  1482 non-null   float64
 10  regular_price_pln                  146 non-null    float64
 11  lowest_price_last_30_days_pln      433 non-null    float64
 1

Dane o produktach (DataFrame products) posiadają 865 ocen produktów a nowe dane (DataFrame reviews) 856. Przeba więc przyjrzeć się tej różnicy.

In [102]:
# Liczba produktów, których url nie znajduje się na liście reviews, i które nie są puste
len(products[
    (~products['url'].isin(reviews['product_url'].unique())) &
    (~products['rating'].isna())
])

35

Takich produktów jest 35, więc na liście muszą znajdować się też produkty, które znajdują się w DataFrame reviews a w DataFrame products mają pusty rating.

In [103]:
len(products[
    (products['url'].isin(reviews['product_url'].unique())) &
    (products['rating'].isna())
])

26

In [104]:
products[
    (products['url'].isin(reviews['product_url'].unique())) &
    (products['rating'].isna())
][['rating', 'url']]

rating  \
0        NaN   
101      NaN   
128      NaN   
130      NaN   
139      NaN   
173      NaN   
284      NaN   
330      NaN   
360      NaN   
380      NaN   
637      NaN   
684      NaN   
700      NaN   
1230     NaN   
1310     NaN   
1358     NaN   
1370     NaN   
1373     NaN   
1385     NaN   
1457     NaN   
1498     NaN   
1527     NaN   
1693     NaN   
1738     NaN   
1753     NaN   
1767     NaN   

                                                                                                                                                                    url  
0                                                https://www.rossmann.pl/Produkt/Balsamy-i-olejki-do-brody/Rituals-Homme-balsam-do-stylizacji-brody-30-ml,2106072,13086  
101                                                    https://www.rossmann.pl/Produkt/Wody-toaletowe/La-Rive-Ironstone-woda-toaletowa-dla-mezczyzn-100-ml,368509,13274  
128   https://www.rossmann.pl/Produkt/Pielegnacja-twarzy/Janda-Men-krem-do-twarzy-suplementacyjny-z-kolagenem-odzywczy-wygladzajacy-przeciwzmarszczkowy-n,2091819,13226  
130                                    https://www.rossmann.pl/Produkt/Wody-toaletowe/Adidas-UEFA-Champions-League-Goal-woda-toaletowa-dla-mezczyzn-50-ml,2085555,13260  
139         https://www.rossmann.pl/Produkt/Pielegnacja-twarzy/LOreal-Paris-Men-Expert-Derma-Control-krem-do-twarzy-dla-mezczyzn-nawilzajacy-SPF-30-40-ml,2100202,13226  
173                                                   https://www.rossmann.pl/Produkt/Wody-perfumowane/Rasasi-Rumz-Al-woda-perfumowana-dla-mezczyzn-50-ml,2109052,13259  
284                                                  https://www.rossmann.pl/Produkt/Wody-toaletowe/Bruno-Banani-Pure-Man-woda-toaletowa-dla-mezczyzn-30-ml,39560,13274  
330                                                 https://www.rossmann.pl/Produkt/Wody-toaletowe/Mexx-Summer-Daydream-woda-toaletowa-dla-mezczyzn-50-ml,2098497,13260  
360                                                            https://www.rossmann.pl/Produkt/Wody-toaletowe/Jaguar-Era-woda-toaletowa-dla-mezczyzn-60-ml,406691,13274  
380                     https://www.rossmann.pl/Produkt/Dezodoranty-do-ciala/Arganove-Active-dezodorant-meski-w-kulce-mineralny-alunowy-Classic-Musk-50-ml,420001,19836  
637                                                                  https://www.rossmann.pl/Produkt/Dezodoranty/Alterra-Men-dezodorant-w-sprayu-24h-75-ml,368081,13119  
684                                                        https://www.rossmann.pl/Produkt/Wody-toaletowe/Hugo-Boss-Man-woda-toaletowa-dla-mezczyzn-100-ml,189503,13274  
700                                         https://www.rossmann.pl/Produkt/Zele-pod-prysznic/Phlov-Mencare-zel-pod-prysznic-do-ciala-i-wlosow-2w1-200-ml,2103479,13229  
1230                                           https://www.rossmann.pl/Produkt/Wody-perfumowane/Artiscent-Santal-Frais-woda-perfumowana-dla-mezczyzn-50-ml,389133,13273  
1310                                                       https://www.rossmann.pl/Produkt/Ekstrakty-perfum/Hugo-Boss-Bottled-perfumy-dla-mezczyzn-100-ml,2092162,19832  
1358                                       https://www.rossmann.pl/Produkt/Balsamy-i-zele-po-goleniu/Masveri-Men-Ice-Up-balsam-po-goleniu-chlodzacy-75-ml,2084870,13243  
1370                                  https://www.rossmann.pl/Produkt/Pielegnacja-zarostu/Rituals-Homme-olejek-do-brody-Olejek-arganowy-i-Bazyliowy-30-ml,2100676,13248  
1373                                                https://www.rossmann.pl/Produkt/Wody-toaletowe/Bruno-Banani-Magic-Man-woda-toaletowa-dla-mezczyzn-30-ml,60412,13274  
1385                                                https://www.rossmann.pl/Produkt/Wody-toaletowe/Lacoste-L1212-Blanc-woda-toaletowa-dla-mezczyzn-100-ml,2077350,13274  
1457                                              https://www.rossmann.pl/Produkt/Zele-pod-prysznic/Rituals-Homme-pianka-pod-prysznic-dla-mezczyzn-200-ml,2096271,13229  
149

Ponieważ mamy dane o ratingu w DF reviews dotyczące powyższych wpisów, uzupełnię te wpisy o dane z DF reviews.

In [105]:
missing_rating_url_list = list(
    products[
        (products['url'].isin(reviews['product_url'].unique())) &
        (products['rating'].isna())
    ]['url']
)

rating_from_reviews = reviews.groupby('product_url')['rating'].mean().reset_index()
rating_from_reviews[rating_from_reviews['product_url'].isin(missing_rating_url_list)]



product_url  \
119                                             https://www.rossmann.pl/Produkt/Balsamy-i-olejki-do-brody/Rituals-Homme-balsam-do-stylizacji-brody-30-ml,2106072,13086   
128                                       https://www.rossmann.pl/Produkt/Balsamy-i-zele-po-goleniu/Masveri-Men-Ice-Up-balsam-po-goleniu-chlodzacy-75-ml,2084870,13243   
140                    https://www.rossmann.pl/Produkt/Dezodoranty-do-ciala/Arganove-Active-dezodorant-meski-w-kulce-mineralny-alunowy-Classic-Musk-50-ml,420001,19836   
141                       https://www.rossmann.pl/Produkt/Dezodoranty-do-ciala/Arganove-Active-dezodorant-meski-w-kulce-mineralny-alunowy-Urban-Men-50-ml,420000,19836   
202                                                                 https://www.rossmann.pl/Produkt/Dezodoranty/Alterra-Men-dezodorant-w-sprayu-24h-75-ml,368081,13119   
243                                                       https://www.rossmann.pl/Produkt/Ekstrakty-perfum/Hugo-Boss-Bottled-perfumy-dla-mezczyzn-100-ml,2092162,19832   
352  https://www.rossmann.pl/Produkt/Pielegnacja-twarzy/Janda-Men-krem-do-twarzy-suplementacyjny-z-kolagenem-odzywczy-wygladzajacy-przeciwzmarszczkowy-n,2091819,13226   
355        https://www.rossmann.pl/Produkt/Pielegnacja-twarzy/LOreal-Paris-Men-Expert-Derma-Control-krem-do-twarzy-dla-mezczyzn-nawilzajacy-SPF-30-40-ml,2100202,13226   
388                                  https://www.rossmann.pl/Produkt/Pielegnacja-zarostu/Rituals-Homme-olejek-do-brody-Olejek-arganowy-i-Bazyliowy-30-ml,2100676,13248   
464                                           https://www.rossmann.pl/Produkt/Wody-perfumowane/Artiscent-Santal-Frais-woda-perfumowana-dla-mezczyzn-50-ml,389133,13273   
509                                                  https://www.rossmann.pl/Produkt/Wody-perfumowane/Rasasi-Rumz-Al-woda-perfumowana-dla-mezczyzn-50-ml,2109052,13259   
538                                  https://www.rossmann.pl/Produkt/Wody-toaletowe/Adidas-UEFA-Champions-League-Goal-woda-toaletowa-dla-mezczyzn-100-ml,2085556,13260   
539                                   https://www.rossmann.pl/Produkt/Wody-toaletowe/Adidas-UEFA-Champions-League-Goal-woda-toaletowa-dla-mezczyzn-50-ml,2085555,13260   
542                                                https://www.rossmann.pl/Produkt/Wody-toaletowe/Adidas-Victory-League-woda-toaletowa-dla-mezczyzn-50-ml,409263,13260   
567                                                https://www.rossmann.pl/Produkt/Wody-toaletowe/Bruno-Banani-Magic-Man-woda-toaletowa-dla-mezczyzn-30-ml,60412,13274   
575                                                 https://www.rossmann.pl/Produkt/Wody-toaletowe/Bruno-Banani-Pure-Man-woda-toaletowa-dla-mezczyzn-30-ml,39560,13274   
635                                                       https://www.rossmann.pl/Produkt/Wody-toaletowe/Hugo-Boss-Man-woda-toaletowa-dla-mezczyzn-100-ml,189503,13274   
655                                                           https://www.rossmann.pl/Produkt/Wody-toaletowe/Jaguar-Era-woda-toaletowa-dla-mezczyzn-60-ml,406691,13274   
672                                                   https://www.rossmann.pl/Produkt/Wody-toaletowe/La-Rive-Ironstone-woda-toaletowa-dla-mezczyzn-100-ml,368509,13274   
676                                                https://www.rossmann.pl/Produkt/Wody-toaletowe/Lacoste-L1212-Blanc-woda-toaletowa-dla-mezczyzn-100-ml,2077350,13274   
689                                                https://www.rossmann.pl/Produkt/Wody-toaletowe/Mexx-Summer-Daydream-woda-toaletowa-dla-mezczyzn-50-ml,2098497,13260   
731                                 https://www.rossmann.pl/Produkt/Wody-toaletowe/The-Master-Perfumer-Marine-Aqua-N45-woda-toaletowa-dla-mezczyzn-100-ml,388665,13260   
733                                https://www.rossmann.pl/Produkt/Wody-toaletowe/The-Master-Perfumer-Urban-Forest-N17-woda-toaletowa-dla-mezczyzn-100-ml,388666,13274   
841                                        https://www.rossmann.pl/Produkt/

Rzeczywiście dane o ratingu dla tych 26 wpisów istnieją. Wystarczy więc przepisać wartości z pomocniczego DF (rating_from_reviews) do DF products.

In [ ]:
merged = products.merge(
    rating_from_reviews[['product_url', 'rating']],
    how = 'left',
    left_on = 'url',
    right_on = 'product_url',
    suffixes = ('', '_from_reviews')
)

merged['rating'] = merged['rating'].combine_first(merged['rating_from_reviews'])
merged = merged.drop(columns=['rating_from_reviews', 'product_url'])

merged[merged['url'].isin(missing_rating_url_list)][['rating', 'url']]


rating  \
0        5.0   
81       5.0   
102      4.0   
104      5.0   
111      5.0   
140      5.0   
224      5.0   
262      5.0   
287      5.0   
304      5.0   
517      4.0   
560      5.0   
574      5.0   
1005     5.0   
1072     5.0   
1115     5.0   
1125     5.0   
1127     5.0   
1137     5.0   
1195     5.0   
1227     5.0   
1250     5.0   
1394     5.0   
1432     5.0   
1444     5.0   
1457     5.0   

                                                                                                                                                                    url  
0                                                https://www.rossmann.pl/Produkt/Balsamy-i-olejki-do-brody/Rituals-Homme-balsam-do-stylizacji-brody-30-ml,2106072,13086  
81                                                     https://www.rossmann.pl/Produkt/Wody-toaletowe/La-Rive-Ironstone-woda-toaletowa-dla-mezczyzn-100-ml,368509,13274  
102   https://www.rossmann.pl/Produkt/Pielegnacja-twarzy/Janda-Men-krem-do-twarzy-suplementacyjny-z-kolagenem-odzywczy-wygladzajacy-przeciwzmarszczkowy-n,2091819,13226  
104                                    https://www.rossmann.pl/Produkt/Wody-toaletowe/Adidas-UEFA-Champions-League-Goal-woda-toaletowa-dla-mezczyzn-50-ml,2085555,13260  
111         https://www.rossmann.pl/Produkt/Pielegnacja-twarzy/LOreal-Paris-Men-Expert-Derma-Control-krem-do-twarzy-dla-mezczyzn-nawilzajacy-SPF-30-40-ml,2100202,13226  
140                                                   https://www.rossmann.pl/Produkt/Wody-perfumowane/Rasasi-Rumz-Al-woda-perfumowana-dla-mezczyzn-50-ml,2109052,13259  
224                                                  https://www.rossmann.pl/Produkt/Wody-toaletowe/Bruno-Banani-Pure-Man-woda-toaletowa-dla-mezczyzn-30-ml,39560,13274  
262                                                 https://www.rossmann.pl/Produkt/Wody-toaletowe/Mexx-Summer-Daydream-woda-toaletowa-dla-mezczyzn-50-ml,2098497,13260  
287                                                            https://www.rossmann.pl/Produkt/Wody-toaletowe/Jaguar-Era-woda-toaletowa-dla-mezczyzn-60-ml,406691,13274  
304                     https://www.rossmann.pl/Produkt/Dezodoranty-do-ciala/Arganove-Active-dezodorant-meski-w-kulce-mineralny-alunowy-Classic-Musk-50-ml,420001,19836  
517                                                                  https://www.rossmann.pl/Produkt/Dezodoranty/Alterra-Men-dezodorant-w-sprayu-24h-75-ml,368081,13119  
560                                                        https://www.rossmann.pl/Produkt/Wody-toaletowe/Hugo-Boss-Man-woda-toaletowa-dla-mezczyzn-100-ml,189503,13274  
574                                         https://www.rossmann.pl/Produkt/Zele-pod-prysznic/Phlov-Mencare-zel-pod-prysznic-do-ciala-i-wlosow-2w1-200-ml,2103479,13229  
1005                                           https://www.rossmann.pl/Produkt/Wody-perfumowane/Artiscent-Santal-Frais-woda-perfumowana-dla-mezczyzn-50-ml,389133,13273  
1072                                                       https://www.rossmann.pl/Produkt/Ekstrakty-perfum/Hugo-Boss-Bottled-perfumy-dla-mezczyzn-100-ml,2092162,19832  
1115                                       https://www.rossmann.pl/Produkt/Balsamy-i-zele-po-goleniu/Masveri-Men-Ice-Up-balsam-po-goleniu-chlodzacy-75-ml,2084870,13243  
1125                                  https://www.rossmann.pl/Produkt/Pielegnacja-zarostu/Rituals-Homme-olejek-do-brody-Olejek-arganowy-i-Bazyliowy-30-ml,2100676,13248  
1127                                                https://www.rossmann.pl/Produkt/Wody-toaletowe/Bruno-Banani-Magic-Man-woda-toaletowa-dla-mezczyzn-30-ml,60412,13274  
1137                                                https://www.rossmann.pl/Produkt/Wody-toaletowe/Lacoste-L1212-Blanc-woda-toaletowa-dla-mezczyzn-100-ml,2077350,13274  
1195                                              https://www.rossmann.pl/Produkt/Zele-pod-prysznic/Rituals-Homme-pianka-pod-prysznic-dla-mezczyzn-200-ml,2096271,13229  
122

In [ ]:
products = merged.reset_index(drop = True)
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1482 entries, 0 to 1481
Data columns (total 54 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   product_id                         1482 non-null   int64  
 1   name                               1482 non-null   string 
 2   caption                            1482 non-null   string 
 3   cat_1                              1482 non-null   string 
 4   cat_2                              1482 non-null   string 
 5   cat_3                              1482 non-null   string 
 6   cat_4                              382 non-null    string 
 7   cat_5                              1482 non-null   string 
 8   url                                1482 non-null   object 
 9   current_price_pln                  1482 non-null   float64
 10  regular_price_pln                  146 non-null    float64
 11  lowest_price_last_30_days_pln      433 non-null    float

26 brakujących ratingów zostało uzupełnione. Usunę kolumnę reviews, ponieważ szczegółowe dane przechowuję w DF reviews.


In [108]:
products = products.drop(columns = ['reviews'])
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1482 entries, 0 to 1481
Data columns (total 53 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   product_id                         1482 non-null   int64  
 1   name                               1482 non-null   string 
 2   caption                            1482 non-null   string 
 3   cat_1                              1482 non-null   string 
 4   cat_2                              1482 non-null   string 
 5   cat_3                              1482 non-null   string 
 6   cat_4                              382 non-null    string 
 7   cat_5                              1482 non-null   string 
 8   url                                1482 non-null   object 
 9   current_price_pln                  1482 non-null   float64
 10  regular_price_pln                  146 non-null    float64
 11  lowest_price_last_30_days_pln      433 non-null    float

Mamy więc 856 ratingów unikalnych produktów w DF reviews i 891 w DF products. W DF reviews brakuje więc danych 35 a nie 9 produktów (różnica wynika z tych 26 produktów, które nie miały ratingu w DF products).

In [ ]:
# Liczba produktów, których url nie znajduje się na liście reviews, i które nie są puste w DF products
len(products[
    (~products['url'].isin(reviews['product_url'].unique())) &
    (~products['rating'].isna())
])

35

In [118]:
# Lista URL tych produktów
missing_data_url_list = list(products[
    (~products['url'].isin(reviews['product_url'].unique())) &
    (~products['rating'].isna())
]['url'])
missing_data_url_list

['https://www.rossmann.pl/Produkt/Pielegnacja-twarzy/AA-Men-Advanced-Care-Vital-krem-do-twarzy-przeciwzmarszczkowy-40-50-ml,2100242,13226',
 'https://www.rossmann.pl/Produkt/Dezodoranty/Old-Spice-Bearglove-dezodorant-w-sprayu-150-ml,198171,13119',
 'https://www.rossmann.pl/Produkt/Minikosmetyki/Joanna-Power-Men-zel-pod-prysznic-3w1-mini-100-ml,205771,19247',
 'https://www.rossmann.pl/Produkt/Wody-toaletowe/Calvin-Klein-Euphoria-woda-toaletowa-dla-mezczyzn-100-ml,193190,13274',
 'https://www.rossmann.pl/Produkt/Wody-toaletowe/Bruno-Banani-Absolute-Man-woda-toaletowa-dla-mezczyzn-50-ml,193534,13274',
 'https://www.rossmann.pl/Produkt/Zele-i-pianki-pod-prysznic/Palmolive-Men-zel-pod-prysznic-dla-mezczyzn-3w1-Citrus-Crush-500-ml,265299,13102',
 'https://www.rossmann.pl/Produkt/Zele-pod-prysznic/Adidas-Pure-Game-zel-pod-prysznic-dla-mezczyzn-3w1-Smoky-Woods-400-ml,411084,13229',
 'https://www.rossmann.pl/Produkt/Balsamy-i-zele-po-goleniu/Phlov-Mencare-Energy-Booster-serum-po-goleniu-lagodza

Dane tych produktów trzeba ponownie pobrać ze strony. Ponownie użyję zmodyfikowanego scrapera.

In [121]:
pd.to_pickle(missing_data_url_list, '../data/reviews_data_2.pkl')

In [125]:
reviews_missing = pd.DataFrame(pd.read_pickle('../data/reviews_missing_2.pkl'))
reviews_missing.head()

product_url  \
0  https://www.rossmann.pl/Produkt/Pielegnacja-twarzy/AA-Men-Advanced-Care-Vital-krem-do-twarzy-przeciwzmarszczkowy-40-50-ml,2100242,13226   
1  https://www.rossmann.pl/Produkt/Pielegnacja-twarzy/AA-Men-Advanced-Care-Vital-krem-do-twarzy-przeciwzmarszczkowy-40-50-ml,2100242,13226   
2  https://www.rossmann.pl/Produkt/Pielegnacja-twarzy/AA-Men-Advanced-Care-Vital-krem-do-twarzy-przeciwzmarszczkowy-40-50-ml,2100242,13226   
3  https://www.rossmann.pl/Produkt/Pielegnacja-twarzy/AA-Men-Advanced-Care-Vital-krem-do-twarzy-przeciwzmarszczkowy-40-50-ml,2100242,13226   
4  https://www.rossmann.pl/Produkt/Pielegnacja-twarzy/AA-Men-Advanced-Care-Vital-krem-do-twarzy-przeciwzmarszczkowy-40-50-ml,2100242,13226   

   rating    author        date content  
0     5.0  Zbigniew  26.06.2025          
1     5.0     Marta  18.06.2025          
2     4.0    Tomasz  16.06.2025          
3     5.0   Dominik  15.06.2025          
4     5.0    Michał  15.06.2025

In [130]:
reviews_combined = pd.concat([reviews, reviews_missing], ignore_index=True)
reviews_combined['product_url'].nunique()

891

Dane w obu DF są teraz zgodne - istnieją dane o ocenie użytkowników dla 891 produktów. Do dalszej analizy weźmiemy tylko produkty zawierające oceny użytkowników.

In [136]:
reviews_combined.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10873 entries, 0 to 10872
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   product_url  10873 non-null  object 
 1   rating       10873 non-null  float64
 2   author       10873 non-null  object 
 3   date         10873 non-null  object 
 4   content      10873 non-null  object 
dtypes: float64(1), object(4)
memory usage: 424.9+ KB


Stworzę DF zawierający dane z DF reviews połączone z niektórymi danymi z DF products - id, marką, nazwą produktu, docelową kategorią, aktualną ceną i wartością przeceny. Usunę natomiast product_url. Wynikowo będę miał długi DF, w którym dla każdego komentarza będę miał kluczowe informacje o produkcie.

In [149]:
reviews_extended = reviews_combined.merge(
    products[['product_id', 'name', 'url', 'brand', 'current_price_pln', 'discount_pct', 'top_level_category']],
    how = 'left',
    left_on = 'product_url',
    right_on = 'url'
)
reviews_extended.drop(columns = ['product_url', 'url'], inplace = True)
# Zmiana formatu daty
reviews_extended['date'] = pd.to_datetime(reviews_extended['date'], format='%d.%m.%Y')

reviews_extended.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10873 entries, 0 to 10872
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   rating              10873 non-null  float64       
 1   author              10873 non-null  object        
 2   date                10873 non-null  datetime64[ns]
 3   content             10873 non-null  object        
 4   product_id          10873 non-null  int64         
 5   name                10873 non-null  string        
 6   brand               10873 non-null  object        
 7   current_price_pln   10873 non-null  float64       
 8   discount_pct        1960 non-null   float64       
 9   top_level_category  10873 non-null  string        
dtypes: datetime64[ns](1), float64(3), int64(1), object(3), string(2)
memory usage: 849.6+ KB


In [150]:
reviews_extended.head()

rating      author       date                                    content  \
0     5.0         EwA 2025-06-25  Mąż bardzo zadowolony z jakości i zapachu   
1     5.0       Olena 2025-05-21                                              
2     4.0    Wojciech 2025-05-21                           przyjemny zapach   
3     5.0       Daria 2025-06-22                                              
4     5.0  Aleksandra 2025-06-12                                              

   product_id                       name       brand  current_price_pln  \
0     2106072          RITUALS... Homme   RITUALS...              74.99   
1     2083791        JAGUAR Era Reloaded      JAGUAR             154.99   
2     2083791        JAGUAR Era Reloaded      JAGUAR             154.99   
3      235165  HUGO BOSS Just Different    HUGO BOSS             167.99   
4      235165  HUGO BOSS Just Different    HUGO BOSS             167.99   

   discount_pct         top_level_category  
0           NaN  Balsamy i olejki do brody  
1           NaN          Zestawy zapachowe  
2           NaN          Zestawy zapachowe  
3           NaN             Wody toaletowe  
4           NaN             Wody toaletowe

Przejście do kolejnego etapu wymagało pobrania dodatkowych danych oraz ponownego przejścia procesu czyszczenia, weryfikowania i łączenia danych. W związku z tym postanowiłem podzielić ten etap na dwie części. Właściwą analizę zebranych i wyczyszczonych danych opiszę w kolejnym artykule.

In [151]:
pd.to_pickle(products, '../data/products.pkl')
pd.to_pickle(reviews_extended, '../data/reviews_extended.pkl')